enhancements/installer/assisted-installer-in-cluster.md

```diff
@@ -0,0 +1,580 @@
+1. A new controller, the Baremetal Agent Controller, sees the matching BareMetalHosts and boots them using the discovery ISO URL found in the InstallEnv's status.
+1. The Agent starts up on each host and reports back to the assisted service, which creates an Agent resource in the cluster. The Agent is automatically labeled with the labels that were specified in the InstallEnv's Spec.
+1. The Baremetal Agent Controller sets the Agent's Role field in its spec to "master" or "worker" if a corresponding label was present on its BareMetalHost.
+1. The Agent is automatically marked as Approved via a field in its Spec based on being recognized as running on the known BareMetalHost.
+1. The Agent runs through the validation and inspection phases. The results are shown on the Agent's Status, and eventually a condition marks the Agent as "ready".
+1. Cluster Creator creates a ClusterDeployment describing a new cluster. It describes how many control-plane and worker agents to expect. It also includes a label selector to match Agent resources that should be part of the cluster.
+1. Cluster Creator applies a label to Agents if necessary so that they match the ClusterDeployment's selector.
+1. Once there are enough ready Agents of each role to fulfill the expected number as expressed on the ClusterDeployment, installation begins.
+
+#### Static Networking
+
+Some customers have asked for the ability to provide static network details
+up-front for each host instead of using DHCP. They want to define this
+configuration at the same time they define the corresponding BareMetalHost.
+
+A new resource called NMStateConfig will have a Spec with the following
+fields:
+
+* MACAddress: a MAC address for any network device on the host to which this config should be applied. This value is only used to ensure that the config is applied to the intended host.
+* Config: a byte array that can contain a raw [nmstate](https://www.nmstate.io/) network config.
+
+nmstate is already in use within OpenShift for applying network configuration
+to nodes via the [kubernetes-nmstate
+operator](https://github.com/nmstate/kubernetes-nmstate). It uses a similar
+pattern of embedding a raw nmstate yaml structure as a byte array in the Spec.
+
+Each NMStateConfig resource will have a label that corresponds to a InstallEnv.
+The raw YAML configs for each matching resource will be rendered to a network
+config by the Assisted Service and then embedded into the discovery ISO for
+that InstallEnv. At runtime, the discovery ISO will find the config that
+matches a MAC address on the current host and then apply the config. It does
+not matter which interface has the matching MAC address; the matching is merely
+used to identify that the current host corresponds to a given config.
+
+The NMStateConfig resource design is being discussed [in a
+pull request](https://github.com/openshift/assisted-service/pull/969).
```

Comment:
I think aligning around nmstate makes sense. And I do really like this idea of the ISO containing a bundle of configs mapped by MAC address.

But I think there's a broader conversation required here than can be captured in a small section of this enhancement.

e.g. we need to consider the existing nmstate CRDs, and other work to support static host network configuration, and the need to even have the sort of per-cluster bonding/vlan configuration that we currently only support by modifying our RHCOS bootimages

For example, I can imagine the CoreOS team agreeing that this should be adopted to "embed your network config in our Live ISO in /etc/coreos-network and a built-in service will apply the appropriate config for that machine, and then it will be persisted to by coreos-installer --copy-network" as an automation variant of "boot our Live ISO, run nmtui, and the network config will be persisted by coreos-installer --copy-network".

I strongly recommend moving this into a separate enhancement. Otherwise I don't see how we can be confident the APIs we introduce here for networking config are taking us in the right direction